core/src/main/java/org/jdbi/v3/core/argument/JavaTimeArgumentFactory.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.argument;
+
+import java.lang.reflect.Type;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.sql.Types;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZonedDateTime;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.jdbi.v3.core.config.ConfigRegistry;
+
+import static org.jdbi.v3.core.generic.GenericTypes.getErasedType;
+
+class JavaTimeArgumentFactory implements ArgumentFactory {
+    private static final Map<Class<?>, ArgBuilder<?>> BUILDERS = createInternalBuilders();
+
+    private static <T> void register(Map<Class<?>, ArgBuilder<?>> map, Class<T> klass, int type, StatementBinder<T> binder) {
+        map.put(klass, (ArgBuilder<T>) v -> new BuiltInArgument<>(klass, type, binder, v));
+    }
+
+    private static Map<Class<?>, ArgBuilder<?>> createInternalBuilders() {
+        final Map<Class<?>, ArgBuilder<?>> map = new IdentityHashMap<>();
+
+        register(map, Instant.class, Types.TIMESTAMP, (p, i, v) -> p.setTimestamp(i, Timestamp.from(v)));
+        register(map, LocalDate.class, Types.DATE, (p, i, v) -> p.setDate(i, java.sql.Date.valueOf(v)));
+        register(map, LocalTime.class, Types.TIME, (p, i, v) -> p.setTime(i, Time.valueOf(v)));
+        register(map, LocalDateTime.class, Types.TIMESTAMP, (p, i, v) -> p.setTimestamp(i, Timestamp.valueOf(v)));
+        register(map, OffsetDateTime.class, Types.TIMESTAMP, (p, i, v) -> p.setTimestamp(i, Timestamp.from(v.toInstant())));
+        register(map, ZonedDateTime.class, Types.TIMESTAMP, (p, i, v) -> p.setTimestamp(i, Timestamp.from(v.toInstant())));
+
+        return Collections.unmodifiableMap(map);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Optional<Argument> build(Type expectedType, Object value, ConfigRegistry config) {
+        Class<?> expectedClass = getErasedType(expectedType);
+
+        if (value != null && expectedClass == Object.class) {
+            expectedClass = value.getClass();
+        }
+
+        @SuppressWarnings("rawtypes")
+        ArgBuilder v = BUILDERS.get(expectedClass);
+
+        return Optional.ofNullable(v).map(f -> f.build(value));
```

Comment:
This body gets repeated a lot. We should try to reduce this duplication.

We already have `AbstractArgumentFactory`. Maybe we need a simpler one that doesn't need subclassing, where you pass in a Java type, a SQL type, and a lambda that does the binding. If it's natural enough it might warrant it's own method in `Configurable`.

Reply:
Yeah, I noticed that... The same thing is happening with the collectors atm. I'll try to find a way.